ENGAGEHF/Managers/NotificationManager/NotificationManager.swift

```diff
@@ -0,0 +1,76 @@
+//
+// This source file is part of the ENGAGE-HF project based on the Stanford Spezi Template Application project
+//
+// SPDX-FileCopyrightText: 2023 Stanford University
+//
+// SPDX-License-Identifier: MIT
+//
+
+import FirebaseFunctions
+import Foundation
+import OSLog
+import Spezi
+import SwiftUI
+import UserNotifications
+
+
+class NotificationManager: Module, NotificationHandler, NotificationTokenHandler, EnvironmentAccessible {
+    @Application(\.registerRemoteNotifications) private var registerRemoteNotifications
+    @Dependency(NavigationManager.self) private var navigationManager
+    
+    private let logger = Logger(subsystem: "ENGAGEHF", category: "NotificationManager")
+    
+    
+    func configure() {}
+    
+    
+    func handleNotificationAction(_ response: UNNotificationResponse) async {
+        /// The server should store the action payload to be accessed here. For example:
+        /// {
+        ///     "Simulator Target Bundle": "edu.stanford.bdh.engagehf",
+        ///     "aps": {
+        ///         "alert": {
+        ///             "title": "Medication Uptitration",
+        ///             "body": "There has been a change in your medications."
+        ///        }
+        ///     },
+        ///     "action": "medications"
+        /// }
+        let payload = response.notification.request.content.userInfo["action"] as? String
+        await _ = navigationManager.execute(MessageAction(from: payload))
+    }
+    
+    func handleNotificationsAllowed() async throws {
+        let granted = try await UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .badge, .sound])
+        
+        guard granted else {
+            return
+        }
+        
+#if !targetEnvironment(simulator)
+        let deviceToken = try await registerRemoteNotifications()
+#else
+        let deviceToken = Data()
+#endif
+        self.configureRemoteNotifications(using: deviceToken)
+    }
+    
+    func receiveUpdatedDeviceToken(_ deviceToken: Data) {
+        self.configureRemoteNotifications(using: deviceToken)
+    }
+    
+    
+    private func configureRemoteNotifications(using deviceToken: Data) {
```

Comment:
I am wondering if we should make this function async as well as we can call it from `handleNotificationsAllowed` in an async context and probably just have to create a Task for `receiveUpdatedDeviceToken`.

We should most likely still surface a user error up to the calling function but might want to discard it in a log statement in `receiveUpdatedDeviceToken`.